File: srcs/server/agents/builtin/llm_ollama.go
package builtin

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// OllamaClient implements LLMClient for Ollama (local LLM).
type OllamaClient struct {
	Endpoint string
	Client   *http.Client
}

func NewOllamaClient(endpoint string) *OllamaClient {
	if endpoint == "" {
		endpoint = "http://localhost:11434/api/chat"
	}
	return &OllamaClient{
		Endpoint: endpoint,
		Client:   &http.Client{Timeout: 5 * time.Minute},
	}
}

func (c *OllamaClient) Chat(ctx context.Context, req ChatRequest) (ChatResponse, error) {
	type ollamaMessage struct {
		Role    string `json:"role"`
		Content string `json:"content"`
	}

	var messages []ollamaMessage
	if req.System != "" {
		messages = append(messages, ollamaMessage{Role: "system", Content: req.System})
	}

	for _, m := range req.Messages {
		messages = append(messages, ollamaMessage{
			Role:    string(m.Role),
			Content: m.Content,
		})
	}

	payload := map[string]interface{}{
		"model":    req.Model,
		"messages": messages,
		"stream":   false,
	}

	body, _ := json.Marshal(payload)

	httpReq, err := http.NewRequestWithContext(ctx, "POST", c.Endpoint, bytes.NewReader(body))
	if err != nil {
		return ChatResponse{}, err
	}

	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.Client.Do(httpReq)
	if err != nil {
		return ChatResponse{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return ChatResponse{}, fmt.Errorf("ollama api error: %s", resp.Status)
	}

	var result struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return ChatResponse{}, err
	}

	return ChatResponse{
		Message: Message{
			Role:    RoleAssistant,
			Content: result.Message.Content,
		},
	}, nil
}

// ChatStream implements real SSE streaming for Ollama.
func (c *OllamaClient) ChatStream(ctx context.Context, req ChatRequest, chunkChan chan<- ChatResponseChunk) error {
	req.Stream = true
	b, err := json.Marshal(req)
	if err != nil {
		return err
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", c.Endpoint+"/api/chat", bytes.NewReader(b))
	if err != nil {
		return err
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Accept", "text/event-stream")

	resp, err := c.Client.Do(httpReq)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("ollama streaming error: %s", resp.Status)
	}

	scanner := bufio.NewScanner(resp.Body)
	for scanner.Scan() {
		line := scanner.Text()

		var chunk struct {
			Message struct {
				Content   string `json:"content"`
				ToolCalls []struct {
					Function struct {
						Name      string `json:"name,omitempty"`
						Arguments map[string]interface{} `json:"arguments,omitempty"`
					} `json:"function,omitempty"`
				} `json:"tool_calls,omitempty"`
			} `json:"message"`
			DoneReason string `json:"done_reason"`
		}
		if err := json.Unmarshal([]byte(line), &chunk); err == nil {
			var tcs []ToolCall
			for _, t := range chunk.Message.ToolCalls {
				bArgs, _ := json.Marshal(t.Function.Arguments)
				tcs = append(tcs, ToolCall{Name: t.Function.Name, Arguments: bArgs})
			}
			chunkChan <- ChatResponseChunk{
				Delta:      chunk.Message.Content,
				StopReason: chunk.DoneReason,
				ToolCalls:  tcs,
			}
		}
	}
	return scanner.Err()
}

File: srcs/server/agents/builtin/llm_ollama.cc
#include "srcs/server/agents/builtin/llm_ollama.h"

#include <utility>

#include "srcs/server/agents/builtin/http_client.h"
#include "srcs/server/agents/builtin/llm_parsing.h"
#include "absl/base/attributes.h"
#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "nlohmann/json.hpp"

namespace ohc::agent {

OllamaClient::OllamaClient(std::string endpoint)
    : endpoint_(std::move(endpoint)) {}

ABSL_ATTRIBUTE_NOINLINE absl::StatusOr<ChatResponse> OllamaClient::Chat(
  const ChatRequest& req) {
  // ---- Build messages array (OpenAI-compatible) ----------------------------
  nlohmann::json messages = nlohmann::json::array();

  if (!req.system.empty()) {
    messages.push_back({{"role", "system"}, {"content", req.system}});
  }

  for (const auto& m : req.messages) {
    const std::string role(RoleToStringView(m.role));
    nlohmann::json content_val = m.content;

    // Tool results: flatten to text for Ollama which may not support native
    // tool-call formats.
    if (m.role == Role::kTool && !m.tool_results.empty()) {
      std::string combined;
      for (const auto& tr : m.tool_results) {
        if (!combined.empty()) combined += "\n";
        combined += tr.content.empty() ? tr.error : tr.content;
      }
      content_val = std::move(combined);
    }

    messages.push_back({{"role", role}, {"content", content_val}});
  }

  // ---- Build payload -------------------------------------------------------
  nlohmann::json payload;
  payload["model"]    = req.model;
  payload["messages"] = std::move(messages);
  payload["stream"]   = false;

  nlohmann::json options;
  if (req.max_tokens > 0)   options["num_predict"] = req.max_tokens;
  if (req.temperature >= 0) options["temperature"] = req.temperature;
  if (!options.empty())     payload["options"]     = std::move(options);

  // Expose tools if present (Ollama supports tool calls from v0.3+).
  if (!req.tool_definitions.empty()) {
    nlohmann::json tools = nlohmann::json::array();
    for (const auto& td : req.tool_definitions) {
      tools.push_back({
          {"type", "function"},
          {"function", {{"name",        td.name},
                        {"description", td.description},
                        {"parameters",  td.parameters}}},
      });
    }
    payload["tools"] = std::move(tools);
  }

  // ---- HTTP POST -----------------------------------------------------------
  HttpRequest http_req;
  http_req.url     = endpoint_;
  http_req.method  = "POST";
  http_req.body    = payload.dump();
  http_req.headers = {{"Content-Type", "application/json"}};
  // Ollama can be slow on a Pi; give it generous time.
  http_req.timeout_seconds = 600;

  auto resp_or = HttpDo(http_req);
  if (!resp_or.ok()) return resp_or.status();
  if (resp_or->status_code != 200) {
    return absl::InternalError(absl::StrCat(
        "Ollama API HTTP ", resp_or->status_code, ": ", resp_or->body));
  }

  // ---- Parse response ------------------------------------------------------
  return ParseOllamaChatResponse(resp_or->body);
}

}  // namespace ohc::agent
